### ISSUE TITLE
Projectiles Cause Too Much Movement Loss

### ISSUE THREAD

**Issue by Mako88** [dxx-rebirth/dxx-rebirth#159]
In the latest beta, all projectiles (weapons and flares) cause the ship to stop too much. In 58.1 being shot with a flare was hardly noticable in the amount that it slowed your ship. Currently, a flare will stop your ship completely. It seems that all of the weapons have their stopping power scaled up as well.

**Comment by zicodxx** [comment]
Just a quick question: Can you maybe counter-test this with an earlier Rebirth version or even the DOS version via DOSBox? 0.58.1 had quite a few physics alterations I have killed off now and I am not 100% sure if 0.58.1 may be the version misbehaving in this particular case.

**Comment by Mako88** [comment]
I would LOVE to test it in DOS, but I don't know how to get multiplayer working in DOSBox to test this.

**Comment by zicodxx** [comment]
I know it's hard to test that with a flare tho you can get the guidebot shooting at you (by positioning yourself between it and a door). And the other weapons stopping power you could estimate from other bots which don't move around a lot.

**Comment by Mako88** [comment]
I tested it in DOS and it is a bug. (Btw, I figured out how to do multiplayer and it's really not hard).

**Comment by vLKp** [comment]
Mako88 wrote:
> In the latest beta ...

Is this a recent regression?  Bisecting runtime changes is tedious relative to bisecting crashes/hangs.  If you already have a sense of when it broke, that would save me some time.  Identifying the bad commit would be ideal, but even a general hint such as "worked in July, broken in November" would be helpful.

**Comment by zicodxx** [comment]
@Mako88 as @vLKp stated, it may be good to know the time frame on when this issue started to show up. For me it may be good to know if you can even reproduce this issue in a Singleplayer game and if yes, which weapons are affected. I spent the last few hours testing both in Singleplayer and Multiplayer and the issue is hard for me to reproduce. In Singleplayer I cannot reproduce it at all and in Multiplayer I *think* I can observe it when players hit each others with flares. However I had no reliable results as flares and other weapons can stop you easily in all versions if you do not fly full speed - which was even harder to do in Multiplayer. So right now I cannot confirm if this issue even exists as from my Singleplayer based observations, things look normal.

**Comment by zicodxx** [comment]
Okay update:
I spent the whole day testing back and forth between 0.58.1 and the most recent commit. I added a little hack into flare_create that will make the flare spawn right on the opposite end of the room where the player starts in Descent 1 level 1 with this small hack that - in a horrible fashion - converts the second player object into an invisible flare turret:

> Objects[74].pos.x=0; Objects[74].pos.z=9441830; Objects[74].segnum=7; Objects[74].orient.fvec.x=0; Objects[74].orient.fvec.y=0; Objects[74].orient.fvec.z=-F1_0;
> Laser_player_fire( Objects[74], FLARE_ID, 6, 1, Objects[74].orient.fvec);
(In newer revisions, the syntax for this differs a bit)
I also disabled the wiggle to make the results more reliable. So I flew forward and launched a flare at my own ship soon as I reached full speed.

From all I can tell, whether a Flare stops you or not is up to chance. If feel older revisions (like 36cd5404d4d2c8309edee4ca57c9c1f01f003464) seem to give my around 3 out of 10 cases where the ship is completely stopped while on newer revisions it feels more like 7 out of 10. Then again I can test with older revsions, firing at my ship 20 times and get stopped all the time. 

TL;DR: It's hard to nail this down and I have yet to find out *why* this is related to chance other than slight inaccuracies in the physics/collisions code (i.e. how far the flare intersects with the player object until the collision is triggered). It does not seem to change with varying FPS - which is good - but whether a flare will stop you or not is hard to predict.

**Comment by zicodxx** [comment]
Finally identified the commit responsible. With the hack from my last comment and using as many flares as possible to hit myself, I could easily identify which revisions worked properly and which would not. The commit responsible is: 707b69220d24502bd22a9aabc347e94bc48c746b.

**Comment by Mako88** [comment]
Awesome! Once that commit is reverted (or re-written), I'll recompile the beta and test :)

**Comment by vLKp** [comment]
Why would that commit affect movement loss?  It only controls whether you get hit.  Could you provide observed inputs for the function that produce different results for the old vs new versions?

**Comment by zicodxx** [comment]
Okay so with the code snippet posted above - if you put that in flare_create, you can launch Descent 1 level 1 and have flares fired right at you. If you fly forward while firing as many flares as possible, you should reach the opposite wall (where flares originate from) rather fast - usually before your shields drop below 86. With the commit  707b692 however, you will be stopped so much by the flares that by the time you reach the wall, you will have your shields dropped below 75 approximately. 

What I think is that the change you made makes the collision detection much more accurate. It is easily visible that with this commit, all flares hitting me cause a visible spark on my screen while before that, some flares seem to intersect with the Pyro further (as sometimes the hit sparks are just barely visible on the screen edges or not visible at all) causing the movement loss having much less impact. 

At least this is my theory. However this commit definitely seem to have introduced this change. 
If it turns out to be true and a desirable change, then I would propose I add a slight change. There is a piece of code in do_physics_sim that changes an object velocity of an object after a collision. With a certain randomness for weapon collisions, it should easily be possible to recreate the original game behavior - without forcing inaccurate fvi code.

**Comment by vLKp** [comment]
That is surprising.  It should not improve accuracy, either.  It was written as a performance improvement to skip computing sqrt when no collision is possible.  The code for reducing movement does not seem to care where you got hit, only whether you did:
```c++

	// After collision with objects and walls, set velocity from actual movement
	if (!obj_stopped && !bounced 
		&& ((obj->type == OBJ_PLAYER) || (obj->type == OBJ_ROBOT) || (obj->type == OBJ_DEBRIS)) 
		&& ((fate == HIT_WALL) || (fate == HIT_OBJECT) || (fate == HIT_BAD_P0))
		)
	{	
		const auto moved_vec = vm_vec_sub(obj->pos,start_pos);
		vm_vec_copy_scale(obj->mtype.phys_info.velocity,moved_vec,fixdiv(f1_0,FrameTime));
	}
```
If `obj->pos` is updated only by `do_physics_sim`, then none of the terms vary their intensity by the type of impact or its location.

**Comment by zicodxx** [comment]
Yeah this code indeed doesn't care but I could use it to compensate for the change. Tho if you say that the fvi change would not improve accuracy or change anything else, then it may be worth investigating further. I can definitely confirm the change comes from there and I can also supply you with demos tomorrow demonstrating the difference. I will also add more output and try and see why exactly it behaves differently.

**Comment by zicodxx** [comment]
Okay I recorded a few test demos which you can download here: http://dxx-rebirth.com/tmp/compdem.tar.gz
These are D1X demos from Level 1. GOOD.dem is the demo made with the commit immediately before 707b692 and BROKEN.dem is made with 707b692. In the latter you can see how much more force is applied by the constant rain of flares.

I also added some more output - in fact I compiled the code so it produces the variables from both before and after the commit. 

The change from "if (dist < sphere_rad)" to "if (dist2 < sphere_rad_squared)" does not seem to affect the changed behavior at all. But I noticed that the variable shorten seems to have changed drastically - mostly (not always), when a collision with a flare happens, it's 0.
shorten before / after 707b692
326656 / 0
326656 / 0
301568 / 7338240
326656 / 0
301056 / 5099264
326656 / 0
326656 / 0

Could it be that we may need a fix64 version of fix_sqrt?

**Comment by vLKp** [comment]
We already have a `quad_sqrt` used by the vector math library, so adding an overload to convert a fix64 to a quadint is easy.
```diff
diff --git c/common/include/maths.h w/common/include/maths.h
index 2394d2d..ec6340e 100644
--- c/common/include/maths.h
+++ w/common/include/maths.h
@@ -149,6 +149,9 @@ uint32_t quad_sqrt (quadint);
 __attribute_warn_unused_result
 fix fix_sqrt (fix a);
 
+__attribute_warn_unused_result
+fix fix_sqrt (fix64 a);
+
 struct fix_sincos_result
 {
 	fix sin, cos;
diff --git c/common/maths/fixc.cpp w/common/maths/fixc.cpp
index b661371..5baf804 100644
--- c/common/maths/fixc.cpp
+++ w/common/maths/fixc.cpp
@@ -217,6 +217,13 @@ fix fix_sqrt(fix a)
 	return static_cast<fix>(long_sqrt(a)) << 8;
 }
 
+fix fix_sqrt(fix64 a)
+{
+	quadint q;
+	q.q = a;
+	return static_cast<fix>(quad_sqrt(q)) << 8;
+}
+
 static fix fix_sin(const fix_sincos_input sci)
 {
 	fix ss = sincos_table[sci.m_idx];
```
I copied the `<< 8` from `fix_sqrt(fix)`.  I have not tested this.

**Comment by zicodxx** [comment]
You gotta excuse me but the math is WAY over my head here. However fix_sqrt doesn't seem to be the direct issue but more likely that dist2 and sphere_rad_squared are completely different from fixmul(dist,dist) and fixmul(sphere_rad,sphere_rad) respectively. 
Example from a random collision:
fixmul(dist,dist) =  405739
dist2 = 820663481
and
fixmul(sphere_rad,sphere_rad) = 1733275
sphere_rad_squared = 1922767460

Because of this the result of "shorten" is completely different and as a result, using it with w_dist, this completely changes int_dist, too.

**Comment by vLKp** [comment]
How did you print out those values?  They look like they were truncated at 32 bits.

I pushed [Fix fvi sqrt](/dxx-rebirth/dxx-rebirth/commit/0c0dff9def7e06fd1ab032ef41b3013ce343c6ac) to add a missing `>> 16` to the computation.  `fixmul` is not a straight wrapper for `operator*(int,int)`, but earlier work acted like it was.  If you still see problems, then the full output of **info locals** and **info args** when stopped at a gdb prompt after assigning `int_dist` in a bad collision would be helpful.  You will probably need to switch to `-O0` to prevent the compiler from optimizing out some of the important temporaries.

**Comment by zicodxx** [comment]
Admittedly, maybe I didn't cast the values right but this is how I did it:
> const auto dist = vm_vec_dist(closest_point,sphere_pos);                
> printf("dist2old %i dist2new %li sradold %i sradnew %li\n",fixmul(dist,dist), static_cast<fix64>(dist2), fixmul(sphere_rad,sphere_rad), static_cast<fix64>(sphere_rad_squared));

The good news is, your recent commit 0c0dff9def7e06fd1ab032ef41b3013ce343c6ac does in fact fix the issue. I ran several tests with Flares and it behaved again just as it should. I also added the old code before the optimization to compute "shorten_old" and it always equals shorten now. Thanks :)
